packages/react-components/react-tree/src/hooks/useNestedTreeNavigation.ts

```diff
@@ -0,0 +1,128 @@
+import { ArrowDown, ArrowLeft, ArrowRight, ArrowUp, End, Home } from '@fluentui/keyboard-keys';
+import { TreeNavigationData_unstable } from '../Tree';
+import { filterTreeItemAndGroup } from '../utils/filterTreeItemAndGroup';
+import { isHTMLElementWithRole } from '../utils/isHTMLElementWithRole';
+import { useDOMTreeWalker } from './useDOMTreeWalker';
+
+export function useNestedTreeNavigation() {
+  const [treeWalkerRef, rootRef] = useDOMTreeWalker(NodeFilter.SHOW_ELEMENT, {
+    acceptNode: filterTreeItemAndGroup,
+  });
+
+  function navigate(data: TreeNavigationData_unstable) {
+    const { current: treeWalker } = treeWalkerRef;
+    if (!treeWalker) {
+      return;
+    }
+    switch (data.type) {
+      case ArrowLeft:
+        return focusSubtreeOwnerItem(data.target, { treeWalker });
+      case ArrowRight:
+        return focusFirstSubtreeItem(data.target, { treeWalker });
+      case 'TypeAhead':
+        return focusTypeAhead(data.target, { key: data.event.key, treeWalker });
+      case End:
+        return focusLastTreeItem({ treeWalker });
+      case Home:
+        return focusFirstTreeItem({ treeWalker });
+      case ArrowDown:
+        return focusNextTreeItem(data.target, { treeWalker });
+      case ArrowUp:
+        return focusPreviousTreeItem(data.target, { treeWalker });
+    }
+  }
+  return [navigate, rootRef] as const;
+}
+
+type TreeWalkerOptions = { treeWalker: TreeWalker };
+
+export function focusTypeAhead(target: HTMLElement, { treeWalker, key }: { key: string } & TreeWalkerOptions) {
+  treeWalker.currentNode = target;
+  const keyToLowerCase = key.toLowerCase();
+  let nextNode: Node | null;
+  while ((nextNode = treeWalker.nextNode()) !== target) {
+    if (nextNode === null) {
+      treeWalker.currentNode = treeWalker.root;
+      continue;
+    }
+    if (
+      isHTMLElementWithRole(nextNode, 'treeitem') &&
+      nextNode.textContent?.trim().charAt(0).toLowerCase() === keyToLowerCase
+    ) {
+      nextNode.focus();
+      return;
+    }
+  }
+}
+
+function focusFirstTreeItem({ treeWalker }: TreeWalkerOptions) {
+  treeWalker.currentNode = treeWalker.root;
+  const nextNode = treeWalker.firstChild();
+  if (!isHTMLElementWithRole(nextNode, 'treeitem')) {
```

Comment:
why do you assert on the `treeitem` role here and not in the flat tree navigation? You can simply just use a node filter in that case